### ISSUE TITLE
[platform-server] bootstrapApplication have memory leaks, the module generate each time server render 

### ISSUE THREAD

**Issue by hiepxanh** [angular/angular#46473]
### Which @angular/* package(s) are the source of the bug?

platform-server

### Is this a regression?

No

### Description

# ngModule version
I clone the Angular Universal repo version from angular:

- checkout master then: `pnpm install`
- `pnpm dev:ssr` to serve
- `nodemon --inspect=localhost:9236 dist/server/main.js` to inspect

Then I start it and make a heap snapshot. it's fine:
first-time load home page:
![image](https://user-images.githubusercontent.com/16970990/175272300-fb9306f2-1c90-42ec-a7d9-729302bc6772.png)
load home page at second time:
![image](https://user-images.githubusercontent.com/16970990/175272485-6b60981c-b889-4b8f-9179-3e99ef2a37ef.png)

load home page at 20th times:
![image](https://user-images.githubusercontent.com/16970990/175272463-3c3d8131-c363-48bf-901e-4e861569b972.png)

everything is fine and no leak.

# Standalone version

I upgrade the universal version of the standalone component, I'm going to make this on the universal repo, but first, it has a memory leak.

- checkout master then: `pnpm install`
- `pnpm dev:ssr` to serve
- `nodemon --inspect=localhost:9236 dist/server/main.js` to inspect

go home page and f5 for 6 times:
![image](https://user-images.githubusercontent.com/16970990/175273653-42ad493b-93df-4216-8d60-0b0910fcceff.png)

go home page and f5 for more 5 times:
![image](https://user-images.githubusercontent.com/16970990/175273755-f871720a-ab46-4c2a-91b7-867dee1bb9df.png)

we get too much module  is not clear and the heap memory go "To infinity and beyond"

### Please provide a link to a minimal reproduction of the bug

https://github.com/hiepxanh/universal-test

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 14.0.2
Node: 16.13.1
Package Manager: pnpm 7.1.5
OS: win32 x64

Angular: 14.0.3
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, platform-server
... router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1400.2
@angular-devkit/build-angular   14.0.2
@angular-devkit/core            14.0.2
@angular-devkit/schematics      14.0.2
@angular/cli                    14.0.2
@nguniversal/builders           14.0.1
@nguniversal/common             14.0.1
@nguniversal/express-engine     14.0.1
@schematics/angular             14.0.2
rxjs                            7.5.5
typescript                      4.7.4
```

### Anything else?

I'm running the standalone SSR version in my production version. Current work around is restart the server :D

**Comment by hiepxanh** [comment]
in my production app, it goes really bad, it takes 1 day only to eat my 1 GB ram.

![image](https://user-images.githubusercontent.com/16970990/175276203-26448b32-0bf9-47c1-8250-2f86c57f4c9c.png)

![image](https://user-images.githubusercontent.com/16970990/175276048-fbf13fc3-fd98-4b12-980e-d1b38223b08f.png)

what is the best solution to config, or is it a bug?

**Comment by hiepxanh** [comment]
- I only make this modification on express engine
![image](https://user-images.githubusercontent.com/16970990/175276416-46b9890f-dbd7-478f-ad42-48f179554e59.png)

- add appId for render options
![image](https://user-images.githubusercontent.com/16970990/175276509-98c78ee1-b364-43be-95ac-e0c391d628ce.png)

**Comment by hiepxanh** [comment]
sorry, I was tempted. It just too smooth. I used jsdom too, but it event eat more ram than this one.

**Comment by alxhub** [comment]
I believe you're right about the leak. Digging into this just a little, it appears that `@angular/platform-server`'s `_render` is relying on destroying the platform to destroy the application, and the platform seems to rely on registration of bootstrapped NgModules, so indeed this might not be correctly destroying the rendered application.

**Comment by hiepxanh** [comment]
Yes, I do, I also put a log while testing but do not see that log trigger on ngOnDestroy of appComponent

**Comment by AndrewKushnir** [comment]
@hiepxanh I've created a PR with a proposed fix, which ensures that the `ApplicationRef.destroy` is called (and subsequently clears references) when a platform is destroyed. I'd like to ask if you could try using a package [from that PR](https://app.circleci.com/pipelines/github/angular/angular/47728/workflows/f3d30c24-04dc-4895-a264-a76b95ab7a88/jobs/1184898/artifacts) and let us know if the memory leaks is gone. You can update the `package.json` in your app like this (and run `npm i`):
```
  "@angular/core": "https://output.circle-artifacts.com/output/job/9387cf04-2cb0-47ad-b868-42121b6a011f/artifacts/0/angular/core-pr46497-91c5f1807c.tgz",
```
Note: this package is for **testing purposes only** and can not be used in production.

Thank you.

**Comment by hiepxanh** [comment]
@AndrewKushnir  yes sir, I'm very exciting, I'll test on this now

**Comment by hiepxanh** [comment]
@AndrewKushnir  thank you, sir. There is no sign of the memory leak anymore. 
![image](https://user-images.githubusercontent.com/16970990/175755621-51ae7402-e1b9-4e6f-909b-01d6ef51041c.png)

After routing about 100 routes, the memory is still stable. everything is fine. I think I can use it now  <3

**Comment by AndrewKushnir** [comment]
@hiepxanh just wanted to let you know that we've released Angular v14.0.4, which contains the memory leak fix. Please try updating to the latest version and let us know if the problem is resolved. Thank you.

**Comment by hiepxanh** [comment]
Thank you that so fast, I'm going to do it now 😍

**Comment by hiepxanh** [comment]
@AndrewKushnir   I already test it, and I see the result is very good. Thank you for your effor
![image](https://user-images.githubusercontent.com/16970990/176560408-95b97abd-24d5-49ec-81fc-ef1dadc312b1.png)

**Comment by AndrewKushnir** [comment]
@hiepxanh great, thanks for the update!
